odoo/addons/base/models/ir_profile.py

```diff
@@ -0,0 +1,105 @@
+from odoo.exceptions import UserError
+from odoo.http import request
+from odoo.tools.profiler import make_session
+from odoo.tools.speedscope import Speedscope
+
+
+class IrProfile(models.Model):
+    _name = 'ir.profile'
+    _description = 'Profiling results'
+    _log_access = False  # avoid useless foreign key on res_user
+    _order = 'session desc, id desc'
+
+    create_date = fields.Datetime('Creation Date')
+
+    session = fields.Char('Session', index=True)
+    name = fields.Char('Description')
+    duration = fields.Float('Duration')
+
+    init_stack_trace = fields.Text('init_stack_trace', prefetch=False)
+
+    sql = fields.Text('Sql', prefetch=False)
+    traces_async = fields.Text('Traces Async', prefetch=False)
+    traces_sync = fields.Text('Traces Sync', prefetch=False)
+
+    speedscope = fields.Binary('Speedscope', compute='_compute_speedscope')
+    speedscope_url = fields.Text('Open', compute='_compute_speedscope_url')
+
+    @api.autovacuum
+    def _gc_profile(self):
+        # remove profiles older than 30 days
+        domain = [('create_date', '<', fields.Datetime.now() - datetime.timedelta(days=30))]
+        return self.sudo().search(domain).unlink()
+
+    def _compute_speedscope(self):
+        for execution in self:
+            sp = Speedscope(init_stack_trace=json.loads(execution.init_stack_trace))
+            if execution.sql:
+                sp.add('sql', json.loads(execution.sql))
+            if execution.traces_async:
+                sp.add('frames', json.loads(execution.traces_async))
+            if execution.traces_sync:
+                sp.add('settrace', json.loads(execution.traces_sync))
+
+            result = json.dumps(sp.add_default().make())
+            execution.speedscope = base64.b64encode(result.encode('utf-8'))
+
+    def _compute_speedscope_url(self):
+        for profile in self:
+            profile.speedscope_url = f'/web/speedscope/{profile.id}'
+
+    def _enabled_until(self):
+        """
+        If the profiling is enabled, return until when it is enabled.
+        Otherwise return ``None``.
+        """
+        limit = self.env['ir.config_parameter'].sudo().get_param('base.profiling_enabled_until', '')
+        return limit if str(fields.Datetime.now()) < limit else None
+
+    @api.model
+    def set_profiling(self, profile=None, collectors=None, params=None):
+        """
+        Enable or disable profiling for the current user.
+
+        :param profile: ``True`` to enable profiling, ``False`` to disable it.
+        :param list collectors: optional list of collectors to use (string)
+        :param dict params: optional parameters set on the profiler object
+        """
+        # Note: parameters are coming from a rpc calls or route param (public user), 
+        # meaning that corresponding session cariables are client-defined.
+        # This allows to activate any profiler, but can be
+        # dangerous handling request.session.profile_collectors/profile_params.
```

Comment:
For the first part, we may want to discuss about that but the initial demand was by @Gorash that needed to optimise website. Public user is a special case when analysing performances.
It was also in AL specs to be able to enable profiling on any user.
This can only be done when profiling is enabled on the database so the risk is quite low. Anyway:
- we thought of adding a limit on the number of samples on a request/session. 
- another option that I'll send you by mp.

Reply:
> It was also in AL specs to be able to enable profiling on any user.

I was thinking less restriction by user and more e.g. restriction by client's IP. Hence asking if we want to allow drive-by profiling, otherwise we could e.g. only allow profiling from local loopback (by default).

Reply:
We can maybe discuss about that after merge with AL if you want to.